### PR TITLE
[FIX] mail: missing notif in systray

### PR DESCRIPTION
**Pull request by zel-odoo**
When handling notifications in Odoo and tagging the user on a SO/ task / whatever, the messaging counter is increased by 1 and the notification appears in the inbox, but not in the systray.

This is because the needaction is wrongly set with self, which is the user that is doing the action, not the user that should be notified.

This commit fixes the issue by correctly setting user when it comes to the personalized notify

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -3170,7 +3170,7 @@ def _notify_thread_by_inbox(self, message, recipients_data, msg_vals=False, **kw
 
             MailMessage = self.env['mail.message']
             messages_format_prepared = MailMessage._message_format_personalized_prepare(
-                message._message_format(msg_vals=msg_vals, for_current_user=True), partner_ids=inbox_pids)
+                message._message_format(msg_vals=msg_vals, for_current_user=False), partner_ids=inbox_pids)
```
Comment (seb-odoo) [inline addons/mail/models/mail_thread.py:3173]: The base method was fine, this should do a with_user instead

Comment (seb-odoo) [reply addons/mail/models/mail_thread.py:3173]: As part of `_message_format_personalize` probably

File: addons/crm/tests/test_crm_lead_convert_mass.py [hunk 2]
```diff
@@ -42,7 +42,7 @@ def test_assignment_salesmen_wteam(self):
         with self.assertQueryCount(user_sales_manager=0):
             test_leads = self.env['crm.lead'].browse(test_leads.ids)
 
-        with self.assertQueryCount(user_sales_manager=583):  # crm 543 / com 545 / ent 583
+        with self.assertQueryCount(user_sales_manager=585):  # crm 543 / com 545 / ent 583
```
Comment (seb-odoo) [inline addons/crm/tests/test_crm_lead_convert_mass.py:45]: What are the extra queries exactly?
Is that one for each notified user?

Can't we reduce them, by preparing fetching all necessary data for all users beforehand?

Comment (zel-odoo) [reply addons/crm/tests/test_crm_lead_convert_mass.py:45]: it comes from the `self.env['res.partner'].browse(partner_id)` but indeed, we got enough data beforehand and I will see if it can stay the same by doing it correctly.

File: addons/mail/models/mail_message.py [hunk 3]
```diff
@@ -1121,7 +1121,8 @@ def _message_format_personalize(self, partner_id, messages_formatted=None, forma
         :return: list of messages_formatted personalized for the partner
         """
         if not messages_formatted:
-            messages_formatted = self._message_format(format_reply=format_reply, msg_vals=msg_vals, for_current_user=True)
+            # sudo: res.partner to format the messages for the partner
+            messages_formatted = self.with_user(self.env['res.partner'].sudo().browse(partner_id).user_ids[0])._message_format(format_reply=format_reply, msg_vals=msg_vals, for_current_user=True)
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/models/mail_message.py:1125]: i think `with_user` should be called here since the `partner_id` is provided, it is not always calling `self`

**Comment by zel-odoo** [comment]
updated query numbers @seb-odoo

File: addons/mail/models/mail_thread.py [hunk 4]
```diff
@@ -3167,15 +3167,17 @@ def _notify_thread_by_inbox(self, message, recipients_data, msg_vals=False, **kw
                 'res_partner_id': pid,
             } for pid in inbox_pids]
             self.env['mail.notification'].sudo().create(notif_create_values)
-
-            MailMessage = self.env['mail.message']
-            messages_format_prepared = MailMessage._message_format_personalized_prepare(
-                message._message_format(msg_vals=msg_vals, for_current_user=True), partner_ids=inbox_pids)
             for partner_id in inbox_pids:
+                # sudo: res.partner to format personalized message
+                partner = self.env['res.partner'].sudo().browse(partner_id).with_prefetch(inbox_pids)
```
Comment (reth-odoo) [inline addons/mail/models/mail_thread.py:3172]: Why not just browse all ids from the getgo and just to partner.id later? seems more idiomatic

File: addons/mail/models/mail_thread.py [hunk 5]
```diff
@@ -3167,15 +3167,17 @@ def _notify_thread_by_inbox(self, message, recipients_data, msg_vals=False, **kw
...
+                users = partner.user_ids
+                internal_users = users - users.filtered('share')
+                main_user = internal_users[0] if len(internal_users) > 0 else users[0] if len(users) > 0 else self.env['res.users']
```
Comment (reth-odoo) [inline addons/mail/models/mail_thread.py:3175]: `internal_users[0] if internal_users else users[0] if users else self.env['res.users']` for short
or even `(internal_users or users or self.env['res.users'])[:1]`

**Review by reth-odoo — COMMENTED** [review]
Hi there :wave: 

Seems to make sense logically but we could probably keep the batched call and edit the formatted message with the user-specific stuff afterwards (extracting "for_current_user" from "message_format"). As it stands it seems like we'll do a search for every notified partner in `_message_format_personalized_prepare`

(also task ids are super cool)

**Comment by zel-odoo** [comment]
> Hi there 👋
> 
> Seems to make sense logically but we could probably keep the batched call and edit the formatted message with the user-specific stuff afterwards (extracting "for_current_user" from "message_format"). As it stands it seems like we'll do a search for every notified partner in `_message_format_personalized_prepare`
> 
> (also task ids are super cool)

Heeey 👋 ,
I was thinking about the same way to handle that but it didn't work well. The reason was if we extracted the `for_current_user` part out of the method, we needed to re-browser and read fields from `message_sudo` and `record` when calling it again, which creates more queries than not extracting it. 
And actually, when sending to a lot of recipients at the same time, there is only one query extra for reading the `user_ids`, which is not possible to avoid with. See test with more than one partner https://github.com/odoo/odoo/pull/168855/files#diff-6966285e759da835e3575ea5487cc9d843abaac6fd254491cf05c10b2aed4cfaR1390
the extra query is `2024-06-18 11:15:53,202 1241098 WARNING odoo odoo.sql_db: [0.131 ms] query: SELECT "res_users"."id", "res_users"."partner_id", "res_users"."active" FROM "res_users" LEFT JOIN "res_partner" AS "res_users__partner_id" ON ("res_users"."partner_id" = "res_users__partner_id"."id") WHERE ("res_users"."partner_id" IN (303)) ORDER BY "res_users__partner_id"."name"  , "res_users"."login"  `

File: addons/mail/models/mail_thread.py [hunk 6]
```diff
@@ -3167,15 +3167,17 @@ def _notify_thread_by_inbox(self, message, recipients_data, msg_vals=False, **kw
                 'res_partner_id': pid,
             } for pid in inbox_pids]
             self.env['mail.notification'].sudo().create(notif_create_values)
+            # sudo: res.partner to format personalized message
+            partners = self.env['res.partner'].sudo().browse(inbox_pids)
+            for partner in partners:
+                users = partner.user_ids
+                internal_users = users - users.filtered('share')
+                main_user = (internal_users or users or self.env['res.users'])[:1]
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:3175]: Why would we have this kind of fallback when only internal users have access to inbox ?

File: addons/mail/models/mail_thread.py [hunk 7]
```diff
@@ -3167,15 +3167,17 @@ def _notify_thread_by_inbox(self, message, recipients_data, msg_vals=False, **kw
...
-            MailMessage = self.env['mail.message']
-            messages_format_prepared = MailMessage._message_format_personalized_prepare(
-                message._message_format(msg_vals=msg_vals, for_current_user=True), partner_ids=inbox_pids)
-            for partner_id in inbox_pids:
                 bus_notifications.append(
-                    (self.env['res.partner'].browse(partner_id),
+                    (partner,
                      'mail.message/inbox',
-                     MailMessage._message_format_personalize(partner_id, messages_format_prepared)[0])
+                     message.with_user(main_user)._message_format_personalize(partner.id, msg_vals=msg_vals))
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:3180]: Why this change ? Why do we need a "with_user", as anyway what matter in personalize if the recipient partner ?

**Review by tde-banana-odoo — CHANGES_REQUESTED** [review]
Some comments, code seems quite convoluted compared to what it tries to fix, especially it has no test (hence: please add a unit test highlighting the issue :) ) .

File: addons/mail/models/mail_thread.py (hunk 6 above)
Comment (zel-odoo) [reply addons/mail/models/mail_thread.py:3175]: the internal part can be removed but there's a case where partner having more than one user_id, so we still need the users[:-1]

File: addons/mail/models/mail_thread.py (hunk 7 above)
Comment (zel-odoo) [reply addons/mail/models/mail_thread.py:3180]: bug reported: when handling notifications in Odoo and tagging the user on a SO/ task / whatever, the messaging counter is increased by 1 and the notification appears in the inbox, but not in the systray.

reason: previously `            messages_format_prepared = MailMessage._message_format_personalized_prepare(
                message._message_format(msg_vals=msg_vals, for_current_user=True), partner_ids=inbox_pids)
 `in `_message_format` `for_current_user=True` but that is the user that is doing the action, not the user that should be notified.

**Comment by zel-odoo** [comment]
Ok, I understand what you mean. let's try I pushed the new code and hopefully, it works better.

File: addons/mail/models/mail_message.py [hunk 8]
```diff
@@ -1133,6 +1138,16 @@ def _message_format_personalize(self, partner_id, messages_formatted=None, forma
                     'is_active': True,
                     'partner': {'id': partner_id, 'type': "partner"},
                 }
+            if not vals.get("needaction"):
+                # sudo: res.partner to format personalized message
+                partner = self.env['res.partner'].sudo().browse(partner_id)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:1143]: All notifications methods in mail are optimized, and here we are going back to a singleton-based check on partners, then users, without cache and anything ... especially we already computed all this data beforehand using optimized SQL (as we already know which recipients have internal users, ...), so I strongly advise rechecking the whole code and avoiding fetching data we already computed.

Comment (tde-banana-odoo) [reply addons/mail/models/mail_message.py:1143]: Sort of propagating "recipients_data" (being in 15 in a big rebase don't have the 17.3 code here, but normally it is called that way).

Comment (zel-odoo) [reply addons/mail/models/mail_message.py:1143]: I understand the problem you are talking about. In the PR, https://github.com/odoo/odoo/pull/162065/files#diff-6039bdf622bf5417d796a812415303924107a9c8b37be05731aeaedd10beab21R1030 this has been removed and so that's why if we want to use `with_user()`, we have to specify the user one by one and it is indeed violating the optimization you had. What do you @seb-odoo think?

**Review by seb-odoo — CHANGES_REQUESTED** [review]

File: addons/mail/models/mail_message.py [hunk 9]
```diff
@@ -1108,7 +1121,7 @@ def _message_format_personalized_prepare(self, messages_formatted, partner_ids=N
             vals['follower_id_by_partner_id'] = followers_by_record_ref.get((vals['model'], vals['res_id']), dict())
         return messages_formatted
 
-    def _message_format_personalize(self, partner_id, messages_formatted=None, format_reply=True, msg_vals=None):
+    def _message_format_personalize(self, partner_id, user_id=None, messages_formatted=None, format_reply=True, msg_vals=None):
```
[the remark this answers is not in the document]
Comment (zel-odoo) [reply addons/mail/models/mail_message.py:1124]: ok, I checked the commit and we did something similar, only if the uid is got from `(r['id'], r['uid']) for r in recipients_data`, it should be easy to backport without adding extra queries. Is your PR coming soon? I guess it will be easier for me to redo this PR when your PR is merged and backported.

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/models/mail_message.py:1124]: I'm not going to backport my PR, you have to take the interesting part and apply them here.

I copied some of your work here to make it, just I went further with cleaning the methods.

I think runbot should be green, there is no query increase.

Note: commit was updated to add missing sudo and convert also mock server: https://github.com/odoo-dev/odoo/commit/5401b3f35c6feaeabe22d65c6c0820b62098a23a

**Comment by seb-odoo** [comment]
Thank you for the investigation here!

I'll fix it in this PR: https://github.com/odoo/odoo/pull/171944
